lib/spack/spack/spec.py

```diff
@@ -1009,11 +1009,14 @@ def copy(self):
         clone.store_by_child = self.store_by_child
 
         # Copy everything from this dict into it.
-        for dspec in itertools.chain.from_iterable(self.values()):
+        for dspec in self.all_edges():
             clone.add(dspec.copy())
 
         return clone
 
+    def all_edges(self):
+        return (d for values in self.values() for d in values)
+
```

Comment:
Is `spec.edges.values` the same as `spec.values`? If not, then there is a bug, AFAICS:

This takes `self.edges.values()` as input:
```py
        for item in sorted(itertools.chain.from_iterable(self.edges.values())):
```
This takes `self.values()` as input:
```py
        for dspec in itertools.chain.from_iterable(self.values()):
```
Thus, it looks to me as if `all_edges` needs to be renames (since it not only deals with `self.edges`) and AFAICS fixed to require an argument, which is either `self.edges` or `self`, depending on the existing call.